File: LeetCode/Java/src/solutions/p0101_SymmetricTree/P0101.java
package solutions.p0101_SymmetricTree;

import auxiliary.TreeNode;

import java.util.Stack;

public class P0101 {
    public boolean isSymmetric(TreeNode root) {
        // base case
        if (root == null) {
            return true;
        }
        /*
         * Solution v1: Recursive
         *
         * Runtime: 1 ms @ (beats) 0%
         * Memory Usage: 39 MB @ (beats) 0%
         */

        // return isSymmetric(root.left, root.right);

        /*
         * Solution v1: Iterative
         *
         * Runtime: 1 ms @ (beats) 0%
         * Memory Usage: 38.1 MB @ (beats) 55.70%
         */
        Stack<TreeNode> leftNodes = new Stack<>();
        Stack<TreeNode> rightNodes = new Stack<>();
        leftNodes.add(root.left);
        rightNodes.add(root.right);

        while (!leftNodes.isEmpty() && !rightNodes.isEmpty()) {
            TreeNode left = leftNodes.pop();
            TreeNode right = rightNodes.pop();
            // left and right are all null, continue
            if (left == null & right == null) {
                continue;
            }
            // either is null while the other is not null, false
            if (left == null || right == null || left.val != right.val) {
                return false;
            }

            // pay attention to the order of adding element
            // put the nodes by pair into the stack
            leftNodes.add(left.left);
            leftNodes.add(left.right);
            rightNodes.add(right.right);
            rightNodes.add(right.left);
        }
        return true;
    }

    public boolean isSymmetric(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }

        if (left == null || right == null || left.val != right.val) {
            return false;
        }
        // pay attention
        // to judge symmetric, need to compare leftmost and rightmost subtrees
        return isSymmetric(left.left, right.right) && isSymmetric(left.right, right.left);
    }
}

File: LeetCode/Python/solutions/p0101.py
import collections
from typing import Optional

from Python.auxiliary.TreeNode import TreeNode


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 41 ms @ (beats) 29.46%
        # Memory Usage: 14.1 MB @ (beats) 14.51%
        #
        # def helper(tree, reverse=False):
        #     vals = []
        #     q = collections.deque()
        #     q.append(tree)
        #
        #     while q:
        #         curr = q.popleft()
        #
        #         if not curr:
        #             vals.append(None)
        #             continue
        #
        #         vals.append(curr.val)
        #
        #         if reverse:
        #             q.append(curr.left)
        #             q.append(curr.right)
        #         else:
        #             q.append(curr.right)
        #             q.append(curr.left)
        #
        #     return vals
        #
        # left = helper(root.left)
        # right = helper(root.right, True)
        #
        # return left == right

        #
        # Solution v2: Recursive
        #
        # Runtime: 40 ms @ (beats) 35.55%
        # Memory Usage: 13.9 MB @ (beats) 90.34%
        #

        # def check_mirror(left, right):
        #     if not left and not right:
        #         return True
        #     elif not left or not right:
        #         return False
        #     else:
        #         if left.val != right.val:
        #             return False
        #
        #         return check_mirror(left.left, right.right) and \
        #             check_mirror(left.right, right.left)
        #
        # return check_mirror(root.left, root.right)

        #
        # Solution v2.1: Recursive
        # Performance enhance
        #
        # Runtime: 33 ms @ (beats) 79.23%
        # Memory Usage: 14 MB @ (beats) 14.51%
        #
        def check_mirror(left, right):
            if not left and not right:
                return True

            if not left or not right:
                return False

            if left.val != right.val:
                return False

            return check_mirror(left.left, right.right) and \
                check_mirror(left.right, right.left)

        return check_mirror(root.left, root.right)